src/base/associative_cache.hh

```diff
@@ -0,0 +1,338 @@
+        }
+    }
+
+  public:
+
+    /**
+     * Empty constructor - need to call init() later with all args
+     */
+    AssociativeCache() {}
+
+    /**
+     * Public constructor
+     * @param name Name of the cache
+     * @param num_entries total number of entries of the container, the number
+     *   of sets can be calculated dividing this balue by the 'assoc' value
+     * @param associativity number of elements in each associative set
+     * @param entry_size Size of each entry
+     * @param repl_policy replacement policy
+     * @param indexing_policy indexing policy
+     */
+    AssociativeCache(const char *name, const size_t num_entries,
+                     const size_t associativity,
+                     BaseReplacementPolicy *repl_policy,
+                     BaseIndexingPolicy *indexing_policy,
+                     Entry const &init_val = Entry())
+        : cache_name(name), numEntries(num_entries),
+          _associativity(associativity),
+          replPolicy(repl_policy),
+          indexingPolicy(indexing_policy),
+          numSets(num_entries / associativity),
+          entries(num_entries, init_val)
+    {
+        initParams(numEntries, _associativity);
+    }
+
+    ~AssociativeCache() {
+        entries.clear();
+    }
+
+    /**
+     * Return the name of the cache
+     * @return name of the cache
+     */
+    std::string name() const { return cache_name; }
+
+    /**
+     * Clear the entries in the cache.
+     */
+    void clear() {
+        for (auto set = 0; set < numSets; set++) {
+            for (auto way = 0; way < _associativity; way++) {
+                auto repl_entry = indexingPolicy->getEntry(set, way);
+                Entry *entry = static_cast<Entry*>(repl_entry);
+                invalidate(entry);
+            }
+        }
+    }
+
+    void init(const char *name, const size_t num_entries,
+              const size_t associativity,
+              BaseReplacementPolicy *_repl_policy,
+              BaseIndexingPolicy *_indexing_policy,
+              Entry const &init_val = Entry())
+    {
+        cache_name           = std::string(name);
+        numEntries           = num_entries;
+        _associativity       = associativity;
+        replPolicy           = _repl_policy;
+        indexingPolicy       = _indexing_policy;
+        numSets              = num_entries / associativity;
+        entries.resize(num_entries, init_val);
+
+        initParams(numEntries, associativity);
+    }
+
+    /**
+     * Does a valid entry exist for the addr
+     * @param addr Addr to check for
+     * @return if a valid entry exists
+     */
+    virtual bool isEntryValid(const Addr addr)
+    {
+        auto const entry = findEntry(addr);
+
+        return entry && entry->isValid();
+    }
+
+    /**
+     * Get the tag for the addr
+     * @param addr Addr to get the tag for
+     * @return Tag for the address
+     */
+    virtual size_t getTag(const Addr addr) const
```

Comment:
The type of a tag is not size_t. It is either an Addr (preferred), or a new Tag type. Also, is this function really needed? I need to check later

Reply:
Updated it to use Addr.